Add error molecule pattern

discodavey opened the pull request.

discodavey committed:
Add error molecule pattern

discodavey committed:
Add complete error version with languages

davidcmoulton committed:
Merge branch 'master' into add-error-molecule-pattern

davidcmoulton commented on src/patterns/molecules/error/error.scss:
```diff
@@ -0,0 +1,3 @@
+.error {
+  text-align: center;
+}
```
New line.

davidcmoulton commented on src/patterns/molecules/error/error.stories.js:
```diff
@@ -0,0 +1,77 @@
+  },
+);
+
+export const ErrorAr = () => error(
```
Missing `dir` and `lang` attributes.

davidcmoulton commented on src/patterns/molecules/error/error.stories.js:
```diff
@@ -0,0 +1,77 @@
+);
+
+export const ErrorJa = () => error(
+  {
```
Missing `lang` attribute.

davidcmoulton reviewed (commented):
Not many styles in this. Is that intentional because they're for spacing that's not ported over yet, or some other reason?

discodavey committed:
Add code review changes

discodavey committed:
formatting code

discodavey committed:
more formatting code

discodavey committed:
Add styles for error module

discodavey commented:
I have left out the mixins that need the logical properties mixin but have put in a TODO for each one.

discodavey committed:
Merge remote-tracking branch 'upstream/master' into add-error-molecule-pattern

discodavey committed:
Add media-query mixin

discodavey committed:
Merge remote-tracking branch 'upstream/master' into add-error-molecule-pattern

discodavey committed:
Add margin, block-spacing

discodavey committed:
amend code

discodavey committed:
Merge remote-tracking branch 'upstream/master' into add-error-molecule-pattern

discodavey committed:
renaming spacing and block-spacing

discodavey committed:
Merge remote-tracking branch 'upstream/master' into add-error-molecule-pattern

discodavey committed:
Amend for allowing grid

discodavey committed:
Remove addon-centered

centering was confusing patterns that center themselves ie. error molecule pattern

discodavey commented on .storybook/config.js:
```diff
@@ -1,6 +1,5 @@
 import { addDecorator, addParameters, configure } from '@storybook/html';
 import { INITIAL_VIEWPORTS } from '@storybook/addon-viewport';
-import centered from '@storybook/addon-centered/html';
```
Removed as confusion to patterns that centered themselves.

davidcmoulton replied:
I've found the way of disabling `centered` on a per-story basis, see https://github.com/libero/storybook/pull/30/files#r349164152.

For this to work, first you'll need to revert the changes in fa8b5e9 (and then afterwards, to fix up your local environment, run `make install && make build` to get the node module back into your local environment).

davidcmoulton commented on src/patterns/molecules/error/error.scss:
```diff
@@ -0,0 +1,45 @@
+}
+
+.error__heading {
+  @include space.block($end: rem.calculate(baseline.for-units(2)));
```
(Out of scope.) Wondering if we should have a mixin for `rem.calculate(baseline.for-units(2))` as it seems a reasonably common use case.

discodavey replied:
Possibly but I always think having mixins within mixins could confuse and how many mixins we want to nest before it's too much?

davidcmoulton replied:
My thinking is one well chosen name could make the code easier to read, rather than having to mentally parse the nested call in the style rules on each occurrence 🤔

davidcmoulton commented on src/patterns/molecules/error/error.scss:
```diff
@@ -0,0 +1,45 @@
+@use "../../../shared-styles/grid";
+@use "../../../shared-styles/media-query";
+@use "../../../shared-styles/rem";
+@use "../../../shared-styles/scale";
```
`scale ` is not used so should be removed.

davidcmoulton commented on src/shared-styles/_size.scss:
```diff
@@ -3,6 +3,16 @@
+@mixin _constrain-inline-size($size, $extreme) {
+  // TODO: Add Logical Properties when tests have been put in place.
+}
+
```
Can be deleted, as now handled by mixins at https://github.com/libero/storybook/blob/master/src/shared-styles/_size.scss#L26 and https://github.com/libero/storybook/blob/master/src/shared-styles/_size.scss#L34.

davidcmoulton commented on src/shared-styles/_spacing.scss:
```diff
@@ -0,0 +1,66 @@
+@use "logical";
```
The whole file `_spacing.scss` can be removed as it's superseded by `_space.scss`.

davidcmoulton commented on src/shared-styles/_validation.scss:
```diff
@@ -0,0 +1,17 @@
+@mixin expect_at_most($value, $maxLength, $message: "Expected at most #{$maxLength} values") {
```
The whole file `_validation.scss` can be removed as it's superseded by `_validate.scss`.

davidcmoulton commented on src/patterns/molecules/error/error.scss:
```diff
@@ -0,0 +1,45 @@
+
+.error {
+  text-align: center;
+  @include size.max-inline-size(grid.$max-inline-size);
```
(Out of scope) `@include size.max-inline-size()` seems a bit clumsy.

Renaming the dimension constraint mixins could give us either:
1.
  - `@include size.min-inline()`
  - `@include size.max-inline()`

or

2.

  - `@include size.inline-min()`
  - `@include size.inline-max()`

Think I prefer 1 as it more closely matches the underlying css property names. Thoughts? (Would be a separate PR.)

discodavey replied:
I think number one is better as it matches CSS styling like max-width or max-height

davidcmoulton commented on src/patterns/molecules/error/error.scss:
```diff
@@ -0,0 +1,45 @@
+  text-align: center;
+  @include size.max-inline-size(grid.$max-inline-size);
+  @include space.margin(auto, inline);
+  @include space.padding(grid.$column_gap, inline);
```
`grid.$column-gap` (hyphen).

davidcmoulton commented on src/patterns/molecules/error/error.scss:
```diff
@@ -0,0 +1,45 @@
+  @include space.margin(auto, inline);
+  @include space.padding(grid.$column_gap, inline);
+  @include space.block($start: rem.calculate(baseline.for-units(12)), $end: rem.calculate(baseline.for-units(12)));
+  @include media-query.mq($from: medium) {
```
(Out of scope.) While I'm on a roll with the naming revisionism... 😅  I don't much like `@include media-query.mq()` either. `@include media.query()` would read better, but not sure how feasible this is with how we're using the `mq` library. Will have a play and see if can get something sensible from it (into a separate PR).

davidcmoulton commented on src/patterns/molecules/error/error.scss:
```diff
@@ -0,0 +1,45 @@
+}
+
+.error__image {
+  @include space.block($end: baseline.for-units(4));
```
Can just be `@include space.block()` as [the `block()` mixin](https://github.com/libero/storybook/blob/master/src/shared-styles/_space.scss#L32) has `baseline.for-units(4)` as its default `$end` value.

davidcmoulton reviewed (changes requested):
Don't think it needs to be addressed now, but noting that we also have an [`error` Sass module](https://github.com/libero/storybook/blob/master/src/shared-styles/_error.scss) in addition to this error pattern. If this causes confusion, I'd prefer to rename this pattern rather than the sass module, as the `error.throw()`formulation for the sass module works well.

davidcmoulton commented on src/patterns/molecules/error/error.stories.js:
```diff
@@ -0,0 +1,86 @@
+
+export default {
+  title: 'Molecules|Error',
+};
```
To disable the `centered` add-on just for the stories in this file, modify the default export like this:
```
export default {
  title: 'Molecules|Error',
  parameters: {
    centered: {
      disable: true,
    },
  },
};
```
(Reference: https://github.com/storybookjs/storybook/pull/7709.)

discodavey committed:
Merge remote-tracking branch 'upstream/master' into add-error-molecule-pattern

discodavey committed:
Remove as been renamed

discodavey committed:
Amend from code review

discodavey committed:
Disable the centered decorator

davidcmoulton committed:
Re-introduce centered Storybook add-on
